### ISSUE TITLE
Tune.py requires "./" for local binaries on Windows. 

### ISSUE THREAD

**Issue by thomasahle** [thomasahle/fastchess#3]
Usually Windows users don't expect this.

See http://talkchess.com/forum3/viewtopic.php?p=809955#p809955

**Comment by fsmosca** [comment]
I have a problem using engines from engine json file. After checking the code, I get this fix which would work now in windows for me, I can now use engine to be tuned from json file and opp engine also from json file.

https://github.com/fsmosca/fastchess/commit/640e185ef59f0e741c8bc2597776dc972995050c

I am not sure if this would work on linux and other windows users especially when they use engines from json file.

Sample command line with json file
`python tune.py "Deuterium v2019.2.37.52" -opp-engine "Deuterium v2019.2.37.53" -conf engines.json`

**Comment by thomasahle** [comment]
That could be a solution. In commit d20b41f4 I'm now passing the PATH variable to Popen, which should allow not using absolute paths when your engine (or python) is in say `/bin`.

This might also fix things for Windows, since it is for some reason common on Windows to have the current directory included in PATH automatically. I don't think we should mimic this behaviour on other systems though, as it could potentially be a security risk.

In conclusion: Can you see if d20b41f4 is enough to fix things on Windows? Otherwise we should add your 640e185 changes, but for Windows only.

**Comment by fsmosca** [comment]
Unfortunately https://github.com/thomasahle/fastchess/commit/d20b41f498ded52ad6fa8c1172a2985b083651fb alone does not work.

**Comment by thomasahle** [comment]
Ok, let's do your solution then, but only for Windows. That is when `if os.name == 'nt':`

**Comment by fsmosca** [comment]
My changes is in https://github.com/fsmosca/fastchess/commit/949240bfad9ab15d1d47d82ef2fa406c41fc059e

**Comment by thomasahle** [comment]
Super. In that version, will it actually work if the engine is installed somewhere else on the system?
I don't know if that's common on windows, but it seems like we might test if it is in the current directory, and only if so join it with the current path.

**Comment by fsmosca** [comment]
It seems like tune.py needs an engine json file.
```
def load_conf(conf):
    if not conf:
        path = pathlib.Path(__file__).parent.parent / 'engines.json'
        if not path.is_file():
            print('Unable to locate engines.json file.')
            return
        return json.load(open(str(path)))
    else:
        return json.load(open(conf))
```
If so then even if the engine.exe is in the same dir with the tune.py, tune.py will not run if engine json file is missing.

What do you mean by "installed somewhere else on the system?"

**Comment by thomasahle** [comment]
I think a common PATH on windows is `C:\`. That is, if you have an executable like `C:\stockfish.exe`, then typing `stockfish` from any directory should start stockfish.

On Unix, if you want to run something in the same directory, you can always write `./stockfish` say.
I have added the option to write `"workingDirectory": "$FILE"` if you want you current directory to be the one that has the `engines.json`.

**Comment by fsmosca** [comment]
> I think a common PATH on windows is `C:\`. That is, if you have an executable like `C:\stockfish.exe`, then typing `stockfish` from any directory should start stockfish.

It would work if the path is registered in the environment variable.

> On Unix, if you want to run something in the same directory, you can always write `./stockfish` say.

Yes I tried linux before and knew about that ./

> I have added the option to write `"workingDirectory": "$FILE"` if you want you current directory to be the one that has the `engines.json`.

That is also good.

Another option is to make -conf a requirement. Let the user handle this personally.
Example:
`-conf c:\engines\engines.json`
or 
`-conf "c:\my engines\engine in tuning.json"`

**Comment by thomasahle** [comment]
Sorry, maybe I'm confused, but I don't see how having stockfish in C:\ would work if we force join with the workingdirectory like in 640e185. Say working directory is C:\playground, then we'd pass C:\playground\stockfish to popen, and I don't think it would then find C:\stockfish, even if C:\ is in the PATH environment variable.
That's why I say we can (on windows) first look in the working directory (like your commit) and then if the engine is not there, pass the command directly to popen so it can look in the PATH.

Would this not work?

**Comment by fsmosca** [comment]
> Sorry, maybe I'm confused, but I don't see how having stockfish in C:\ would work if we force join with the workingdirectory like in [640e185](https://github.com/thomasahle/fastchess/commit/640e185ef59f0e741c8bc2597776dc972995050c). Say working directory is C:\playground, then we'd pass C:\playground\stockfish to popen, and I don't think it would then find C:\stockfish, even if C:\ is in the PATH environment variable.

Look I am only replying regarding what you think about the following statement.

"I think a common PATH on windows is C:\. That is, if you have an executable like C:\stockfish.exe, then typing stockfish from any directory should start stockfish."

You use the word typing .. from any directory, not passing to popen().

> That's why I say we can (on windows) first look in the working directory (like your commit) and then if the engine is not there, pass the command directly to popen so it can look in the PATH.
> 
> Would this not work?

Since tune.py is dependent on engine json file, there is no need to look for where the engine file is, of whether it is in current/working/whatever directory it is. The info to locate/run the engine are already in that json file. If tune.py still cannot locate the engine given the json file, then inform the user that there is something wrong with the json file. Make the user responsible of setting properly the engine json file.

**Comment by thomasahle** [comment]
I think it just comes down to expectations. If I write `"command": "stockfish"` I assume stockfish will be started from wherever it is installed on the system. And in particular when I write `"command": "python engine.py"` I assume python will be found in path and engine.py in the working directory. Just like in a terminal.

I suggest we add the given workingDirectory to the PATH passed to popen on Windows so it fits with their expectations too.

**Comment by fsmosca** [comment]
> I think it just comes down to expectations. If I write `"command": "stockfish"` I assume stockfish will be started from wherever it is installed on the system. 

Talking about windows, that is not entirely true.

Here is what works on windows and popen_uci() that I have known.

Given we are in c:\skopt
### Tuner files
c:\skopt\tune.py
c:\skopt\arena.py

So our current dir is in tuner files dir.

### Engines
eng_dir = 'c:\engines\sf'
eng_file = 'sf.exe'
eng_dir_file = 'c:\engines\sf\sf.exe' 

(1)
`popen_uci(eng_dir_file)`
ok

(2)
`popen_uci(eng_file)`
not ok

(3)
`popen_uci(eng_file, cwd=eng_dir)`
not ok

(4)
`popen_uci(eng_dir_file, cwd=eng_dir)`
ok, the most ideal, better than in (1)

We can make (2) ok by copying the file sf.exe to c:\skopt (our current dir)

> And in particular when I write `"command": "python engine.py"` I assume python will be found in path and engine.py in the working directory. Just like in a terminal.

So far I have not yet tried how engine.py works, I only tried the exe file.

Do you have windows? Can you test on windows? Do you use Linux?

I use windows, and I don't have linux. If you can not try engine.py on windows I will try it and see what works.

**Comment by thomasahle** [comment]
> `popen_uci(eng_file)`
> not ok

Does the following work on Windows?

    popen_uci(eng_file, env={"PATH": eng_dir})

> So far I have not yet tried how engine.py works, I only tried the exe file.

I only meant to point at the difference between the first argument (the command) and the second argument (the file) which are searched for in different places.

> Do you have windows? Can you test on windows? Do you use Linux?

I can only test on Linux and Mac.

On Mac I can say all of these work:

    popen_uci('stockfish') # Uses default path
    popen_uci('stockfish', cwd='/usr/local/bin', env={'PATH':''}) # Looks in cwd
    popen_uci('stockfish', env={'PATH':'/usr/local/bin'}) # Looks in path
    os.chdir('/usr/local/bin'); popen_uci('stockfish', env={'PATH':''}) # Uses environment cwd

While this don't work (given my current working directory is not /usr/local/bin):

    popen_uci('stockfish', env={'PATH':''})

Naively I would expect it to work the same on Windows, only with /usr/local/bin replaced by c:\engines\sf

I don't care much about the relative location of the tune.py and engines.json file. Maybe that's part of our disagreement?

**Comment by fsmosca** [comment]
> > `popen_uci(eng_file)`
> > not ok
> 
> Does the following work on Windows?
> 
> ```
> popen_uci(eng_file, env={"PATH": eng_dir})

It does not work.

> I don't care much about the relative location of the tune.py and engines.json file. Maybe that's part of our disagreement?

Not really.

For windows:
My json file is like this.
```
[
    {
        "command": "Deuterium_v2019.2.37.52.exe",
        "workingDirectory": "C:/chess/engines/Deuterium",
        "name": "Deuterium v2019.2.37.52",
        "protocol": "uci",
    ...
    ...
    }
```
or `"workingDirectory": "C:\\chess\\engines\\Deuterium",` also works.

eng_dir = json['workingDirectory']
eng_file = json['command']
eng_dir_file = pathlib.Path(eng_dir, eng_file)

And this one would work.
`popen_uci(str(eng_dir_file), cwd=eng_dir)`

**Comment by thomasahle** [comment]
It sounds like the following should work for you then?

    if os.name == 'nt':
        wd_cmd = pathlib.Path(args['workingDirectory'], cmd[0])
        if wd_cmd.is_file():
            cmd[0] = str(wd_cmd)

**Comment by fsmosca** [comment]
That should work, but even better is add the cwd value as well.
`popen_uci(str(eng_dir_file), cwd=eng_dir)`

Lets say we are in c:\skopt

(1)
 `popen_uci(str(eng_dir_file))`
When engine outputs file logs (there are engines that auto generate files, like debug logs, games), etc, those logs will be saved in c:\skopt. I don't like that.

(2)
 `popen_uci(str(eng_dir_file), cwd=eng_dir)`
When engine outputs file logs, etc, those logs will be saved in the dir of the engine itself. Say engine dir is in c:\engines\sf. Log files will be saved there in sf folder.

So (2) is better, because engine logs, etc are saved on engine's dir. Our work dir from where we launch the tuning will be safe of engine file clutters.

**Comment by thomasahle** [comment]
> That should work, but even better is add the cwd value as well.
> `popen_uci(str(eng_dir_file), cwd=eng_dir)`

Right, but we are already adding `cwd`. At least if the user has set `args['workingDirectory']`.
`eng_dir` is `args['workingDirectory']`, no?

**Comment by fsmosca** [comment]
No.

**Comment by thomasahle** [comment]
You wrote yourself that eng_dir = json["WorkingDirectory"]. 
Is eng_dir then where the engine.json file is? 
But how can we add that as the working directory to popen, when the user has already specified a custom working directory?

I don't understand why it doesn't work to add the engine.json file directory to PATH? Does PATH not have any effect at all?

Probably path of our misunderstanding is that I have a shared engines.json file for all engines on my computer, while you seem to have a specific one for each engine in the same directory as the engine itself. Is this at least correctly understood?

To make your use case easier, you can always set ["workingdirectory"] = "$FILE" which makes it the directory of the engines.json file.

**Comment by fsmosca** [comment]
Looking at the code, I see that cwd is added.
`popen_args['cwd'] = args['workingDirectory'].replace('$FILE', curdir)`
That one is enough. Sorry my mistake.

**Comment by fsmosca** [comment]
> Probably path of our misunderstanding is that I have a shared engines.json file for all engines on my computer, while you seem to have a specific one for each engine in the same directory as the engine itself. Is this at least correctly understood?

I am only using one engine json file.

**Comment by thomasahle** [comment]
@fsmosca Are you saying we already have enough? In that case I'll just close this issue :-)

**Comment by fsmosca** [comment]
@thomasahle 
Your load_engine() is different from mine.

Yours:
https://github.com/thomasahle/fastchess/blob/4819ad57130673b1883da154955d12fe26a705db/tools/tune.py#L150

Mine:
https://github.com/fsmosca/fastchess/blob/67fad63bcea686d4540e50b1b8286a7f07ab5852/tools/tune.py#L150

Your code does not work with my system.

**Comment by thomasahle** [comment]
Ok. I've added an extra a join with the current working directory as we talked about.
